### PR TITLE
Quartic bond potential

### PR DESCRIPTION
**Pull request by bcjeong1616**
This is a re-opening and re-targetting of the previously closed PR, [Quartic bond potential #82](https://github.com/mphowardlab/azplugins/pull/82). Changes have been made according to first reviews from mphoward and astatt.

### CODE REVIEW

File: src/BondEvaluatorQuartic.h [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+                {
+                Scalar epsilon = lj2 * lj2 / Scalar(4.0) / lj1;
+                force_divr += r2inv * r6inv * (Scalar(12.0) * lj1 * r6inv - Scalar(6.0) * lj2);
+                bond_eng += r6inv * (lj1 * r6inv - lj2) + epsilon;
+                }
+            else
+                {
+                force_divr += 0;
+                bond_eng += 0;
+                }
```
Comment (mphoward) [inline src/BondEvaluatorQuartic.h:158]:
```suggestion
```

File: src/BondEvaluatorQuartic.h [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+                {
+                force_divr
+                    += r6inv * (Scalar(12.0) * lj1 * r6inv - Scalar(6.0) * lj2) / r / (r + delta);
+                bond_eng += r6inv * (lj1 * r6inv - lj2) + epsilon;
+                }
+            else
+                {
+                force_divr += 0;
+                bond_eng += 0;
+                }
```
Comment (mphoward) [inline src/BondEvaluatorQuartic.h:187]:
```suggestion
```

File: src/bond.py [hunk 3]
```diff
@@ -63,3 +63,104 @@ def __init__(self):
...
+              :math:`[\mathrm{length}]`.
+
+            * ``b_1`` (`float`, **required**) - First quartic potential fitting
+              parameter :math:`b_1` :math:`[\mathrm{length}]`.
+
+            * ``b_2`` (`float`, **required**) - Second quartic potential fitting
+              parameter :math:`b_2` :math:`[\mathrm{length}]`.
+
+            * ``U_0`` (`float`, **required**) - Quartic potential energy barrier height
+              :math:`U_0` at :math:`r_0` when :math:`r_0 > 2^{\frac{1}{6}}\sigma`
```
Comment (mphoward) [inline src/bond.py:120]:
```suggestion
              :math:`U_0` at :math:`r_0` when :math:`r_0 > 2^{1/6}\sigma`
```

File: src/bond.py [hunk 4]
```diff
@@ -63,3 +63,104 @@ def __init__(self):
...
+        params (TypeParameter[``bond type``, dict]):
+            The parameters of the quartic bonds for each particle type.
+            The dictionary has the following keys:
+
+            * ``k`` (`float`, **required**) - quartic attractive force strength
+              :math:`[\mathrm{energy/length^4}]`.
+
+            * ``r_0`` (`float`, **required**) - Location of the quartic potential
+              cutoff :math:`r_0`. Intended to be larger than the WCA cutoff,
+              :math:`2^{\frac{1}{6}}\sigma`. When true,
```
Comment (mphoward) [inline src/bond.py:109]:
```suggestion
              :math:`2^{1/6}\sigma`. When true,
```

File: src/bond.py [hunk 5]
```diff
@@ -63,3 +63,104 @@ def __init__(self):
...
+        \end{eqnarray*}
+
+
+    Attributes:
+        params (TypeParameter[``bond type``, dict]):
+            The parameters of the quartic bonds for each particle type.
+            The dictionary has the following keys:
+
+            * ``k`` (`float`, **required**) - quartic attractive force strength
+              :math:`[\mathrm{energy/length^4}]`.
```
Comment (mphoward) [inline src/bond.py:105]:
```suggestion
              :math:`[\mathrm{energy}/\mathrm{length}^4]`.
```

File: src/bond.py [hunk 6]
```diff
@@ -63,3 +63,104 @@ def __init__(self):
...
+
+    Examples::
+
+        quartic = azplugins.bond.Quartic()
+        quartic.params['A-A'] = dict(k=1434.3, r_0=1.5, b_1=-0.7589, b_2=0.0,
+                                    U_0=67.2234, sigma=1, epsilon=1)
+        quartic.params['B-B'] = dict(k=1434.3, r_0=1.5, b_1=-0.7589, b_2=0.0,
+                                    U_0=67.2234, sigma=1, epsilon=1, delta=0.2)
+
+    The parameters for the examples provided above were derived by Tsige and Stevens in
```
Comment (mphoward) [inline src/bond.py:140]: Please use reST to insert this as a hyperlink, and use the proper doi URL (not the ACS one).

File: src/bond.py [hunk 7]
```diff
@@ -63,3 +63,104 @@ def __init__(self):
...
+            * ``Delta`` (`float`, **optional**) - Shift :math:`\Delta`,
+              defaults to zero :math:`[\mathrm{length}]`.
+
+    Examples::
+
+        quartic = azplugins.bond.Quartic()
+        quartic.params['A-A'] = dict(k=1434.3, r_0=1.5, b_1=-0.7589, b_2=0.0,
+                                    U_0=67.2234, sigma=1, epsilon=1)
+        quartic.params['B-B'] = dict(k=1434.3, r_0=1.5, b_1=-0.7589, b_2=0.0,
+                                    U_0=67.2234, sigma=1, epsilon=1, delta=0.2)
```
Comment (mphoward) [inline src/bond.py:138]: Is delta really zero and 0.2 in the paper below? It seems unlikely. If not, please only include the example from the paper. Please also see here for how to group text and code together in an example using the `rubric` directive:

https://github.com/glotzerlab/hoomd-blue/blob/dc715ea70b65a3330d65c6b6bc51de09b8d1742f/hoomd/mpcd/collide.py#L154-L163

File: src/bond.py [hunk 8]
```diff
@@ -63,3 +63,104 @@ def __init__(self):
...
+              :math:`U_0` at :math:`r_0` when :math:`r_0 > 2^{\frac{1}{6}}\sigma`
+              :math:`[\mathrm{energy}]`.
+
+            * ``epsilon`` (`float`, **required**) - Repulsive WCA interaction energy
+              :math:`\varepsilon` :math:`[\mathrm{energy}]`.
+
+            * ``sigma`` (`float`, **required**) - Repulsive WCA interaction size
+              :math:`\sigma` :math:`[\mathrm{length}]`.
+
+            * ``Delta`` (`float`, **optional**) - Shift :math:`\Delta`,
```
Comment (mphoward) [inline src/bond.py:129]:
```suggestion
            * ``delta`` (`float`, **optional**) - Shift :math:`\Delta`,
```

File: src/BondEvaluatorQuartic.h [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+        U_0 = v["U_0"].cast<Scalar>();
+        delta = v["delta"].cast<Scalar>();
+
+        auto sigma(v["sigma"].cast<Scalar>());
+        auto epsilon(v["epsilon"].cast<Scalar>());
+
+        const Scalar sigma_2 = sigma * sigma;
+        const Scalar sigma_4 = sigma_2 * sigma_2;
+        sigma_6 = sigma_2 * sigma_4;
+        epsilon_x_4 = Scalar(4.0) * epsilon;
```
Comment (mphoward) [inline src/BondEvaluatorQuartic.h:51]:
```suggestion
        epsilon_x_4 = Scalar(4.) * epsilon;
```

File: src/BondEvaluatorQuartic.h [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+        {
+        k = _params.k;
+        r_0 = _params.r_0;
+        b_1 = _params.b_1;
+        b_2 = _params.b_2;
+        U_0 = _params.U_0;
+        delta = _params.delta;
+        lj1 = _params.epsilon_x_4 * _params.sigma_6 * _params.sigma_6;
+        lj2 = _params.epsilon_x_4 * _params.sigma_6;
+        epsilon = _params.epsilon_x_4 / Scalar(4.0);
```
Comment (mphoward) [inline src/BondEvaluatorQuartic.h:126]:
```suggestion
        epsilon = _params.epsilon_x_4 / Scalar(4.);
```

File: src/BondEvaluatorQuartic.h [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+        if (r_red < Scalar(0))
+            {
+            force_divr += Scalar(-1) * k * r_red
+                          * (4 * r_red * r_red - 3 * (b_1 + b_2) * r_red + 2 * b_1 * b_2)
+                          / (r_red + r_0 + delta);
+            bond_eng += k * (r_red - b_1) * (r_red - b_2) * r_red * r_red + U_0;
+            }
+        else
+            {
+            force_divr += 0;
```
Comment (mphoward) [inline src/BondEvaluatorQuartic.h:206]:
```suggestion
```

File: src/bond.py [hunk 12]
```diff
@@ -63,3 +63,104 @@ def __init__(self):
...
+    the simulation state with:
+
+    .. math::
+        :nowrap:
+
+        \begin{eqnarray*}
+        U(r) &= k (r - \Delta - r_0 - b_1)(r - \Delta - r_0 - b_2)(r - \Delta - r_0)^2
+                  + U_0 + U_{\rm WCA}(r), & r < r_0 + \Delta \\
+             &= U_0, & r \ge r_0 + \Delta
+        \end{eqnarray*}
```
Comment (mphoward) [inline src/bond.py:81]: This equation is displaying poorly because it's too wide. Please try to reformat it in a way that looks nicer. You may need to build the docs locally to figure out what works best.

File: src/bond.py [hunk 13]
```diff
@@ -63,3 +63,104 @@ def __init__(self):
...
+    where :math:`\vec{r}` is the vector pointing from one particle to the other in the
+    bond. The potential :math:`U_{\rm WCA}(r)` is given by:
+
+    .. math::
+        :nowrap:
+
+        \begin{eqnarray*}
+        U_{\rm WCA}(r) &= 4 \varepsilon \left[ \left( \frac{\sigma}{r-\Delta} \\
+                                                                            \right)^{12}
+                             - \left( \frac{\sigma}{r-\Delta} \right)^{6} \right] \\
```
Comment (mphoward) [inline src/bond.py:92]:
```suggestion
                             - \left( \frac{\sigma}{r-\Delta} \right)^{6} \right]
```

File: src/bond.py [hunk 14]
```diff
@@ -63,3 +63,104 @@ def __init__(self):
...
+                             + \varepsilon,
+                                          & r < 2^{\frac{1}{6}}\sigma + \Delta  \\
+                           &= 0,          & r \ge 2^{\frac{1}{6}}\sigma + \Delta
```
Comment (mphoward) [inline src/bond.py:95]:
```suggestion
                                          & r < 2^{1/6}\sigma + \Delta  \\
                           &= 0,          & r \ge 2^{1/6}\sigma + \Delta
```

File: src/bond.py [hunk 15]
```diff
@@ -63,3 +63,104 @@ def __init__(self):
...
+    .. math::
+        :nowrap:
+
+        \begin{eqnarray*}
+        U(r) &= k (r - \Delta - r_0 - b_1)(r - \Delta - r_0 - b_2)(r - \Delta - r_0)^2
+                  + U_0 + U_{\rm WCA}(r), & r < r_0 + \Delta \\
+             &= U_0, & r \ge r_0 + \Delta
+        \end{eqnarray*}
+
+    where :math:`\vec{r}` is the vector pointing from one particle to the other in the
```
Comment (mphoward) [inline src/bond.py:83]:
```suggestion
    where :math:`r` is the distance from one particle to the other in the
```

File: .pre-commit-config.yaml
Comment (mphoward) [inline .pre-commit-config.yaml:1]: Please revert the changes to this file. If they need to be updated, we should do that in a separate commit.

**Review by mphoward — CHANGES_REQUESTED** [review]
This looks great! I have some additional comments, mostly on the documentation, and then this will be ready for a final review.

File: src/bond.py (hunk 12 above)
Comment (bcjeong1616) [reply src/bond.py]: Sorry about the delay! Took some time to figure out how to build the docs locally. Thank you so much for catching this, as well as all the other feedback!

Comment (mphoward) [reply src/bond.py]: No problem! Let me know when you're ready for another look.

**Review by astatt — APPROVED** [review]
@mphoward I think all comments are addressed - if you have some time could you check this over and merge if you are happy with it?

**Comment by mphoward** [comment]
The code looks good overall! The documentation is still not rendering well, but I can edit this directly.

@bcjeong1616 can you please enable edits by maintainers?

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by mphoward** [comment]
I have a commit ready to push, so let me know when this is done!

**Comment by bcjeong1616** [comment]
Hello Mike,
Sorry for the delayed reply! I've looked into changing the permission you requested, but it appears that this option is not available for me to change. This is apparently due to how github handles cross-organization PRs, limiting the options available. You can see here that this is a problem many others have been facing since 2021.
https://github.com/orgs/community/discussions/5634.

If you aren't sure how to get around this limitation, I'd be happy to make the commit myself if you'd like to send me the files you'd like to change the formatting for. Otherwise, just let me know what you'd like me to do and I'd be happy to do so.

**Review by mphoward — APPROVED** [review]
Thanks for your work on this!
